Source/WebKit/UIProcess/API/Cocoa/_WKWebPushDaemonConnection.mm

```diff
@@ -0,0 +1,157 @@
+#import <WebCore/ExceptionData.h>
+#import <WebCore/PushPermissionState.h>
+#import <wtf/BlockPtr.h>
+#import <wtf/RetainPtr.h>
+#import <wtf/cocoa/VectorCocoa.h>
+
+@implementation _WKWebPushDaemonConnectionConfiguration
+
+- (instancetype)init
+{
+    self = [super init];
+    if (!self)
+        return nil;
+
+#if ENABLE(RELOCATABLE_WEBPUSHD)
+    _machServiceName = @"com.apple.webkit.webpushd.relocatable.service";
+#else
+    _machServiceName = @"com.apple.webkit.webpushd.service";
+#endif
+    return self;
+}
+
+@end
+
+@implementation _WKWebPushDaemonConnection
+
+- (instancetype)initWithConfiguration:(_WKWebPushDaemonConnectionConfiguration *)configuration
+{
+    if (!(self = [super init]))
+        return nil;
+
+    Vector<uint8_t> hostAppAuditToken = makeVector(configuration.hostAppAuditTokenData);
+    API::Object::constructInWrapper<API::WebPushDaemonConnection>(self, configuration.machServiceName, configuration.partition);
+
+    return self;
+}
+
+static _WKWebPushPermissionState toWKPermissionsState(WebCore::PushPermissionState state)
+{
+    switch (state) {
+    case WebCore::PushPermissionState::Denied:
+        return _WKWebPushPermissionStateDenied;
+    case WebCore::PushPermissionState::Granted:
+        return _WKWebPushPermissionStateGranted;
+    case WebCore::PushPermissionState::Prompt:
+        return _WKWebPushPermissionStatePrompt;
+    }
+
+    return _WKWebPushPermissionStateDenied;
+}
+
+- (void)getPushPermissionStateForScope:(NSURL *)scopeURL completionHandler:(void (^)(_WKWebPushPermissionState))completionHandler
+{
+    _connection->getPushPermissionState(scopeURL, [completionHandlerCopy = makeBlockPtr(completionHandler)](auto result) {
+        completionHandlerCopy(toWKPermissionsState(result));
+    });
+}
+
+- (void)requestPushPermissionForScope:(NSURL *)scopeURL completionHandler:(void (^)(BOOL))completionHandler
+{
+    _connection->requestPushPermission(scopeURL, [completionHandlerCopy = makeBlockPtr(completionHandler)] (bool result) {
+        completionHandlerCopy(result);
+    });
+}
+
+- (void)setAppBadge:(NSUInteger *)badge scopeURL:(NSURL *)scopeURL
+{
+    std::optional<uint64_t> badgeValue = badge ? std::optional<uint64_t> { (uint64_t)badge } : std::nullopt;
+    _connection->setAppBadge(scopeURL, badgeValue);
+}
+
+- (void)subscribeToPushServiceForScope:(NSURL *)scopeURL applicationServerKey:(NSData *)applicationServerKey completionHandler:(void (^)(_WKWebPushSubscriptionData *, NSError *))completionHandler
+{
+    auto key = makeVector(applicationServerKey);
+    _connection->subscribeToPushService(scopeURL, WTFMove(key), [completionHandlerCopy = makeBlockPtr(completionHandler)] (auto result) {
+        if (result)
+            return completionHandlerCopy(wrapper(API::WebPushSubscriptionData::create(WTFMove(result.value()))).get(), nil);
+
+        auto error = [NSError errorWithDomain:@"WKErrorDomain" code:WKErrorUnknown userInfo:@{ NSLocalizedDescriptionKey:result.error().message }];
```

Comment:
I think this is okay for now but maybe in the future we should provide more info in the error since these are literally DOMExceptions. For instance, we can define an error code `WKErrorCodeDOMException` and a userInfo keys `WKErrorCodeDOMExceptionNameKey` and `WKErrorCodeDOMExceptionMessageKey` which contains a string representation of the DOMException name and message respectively. Can be deferred to a future task.

Reply:
Sure; I left a FIXME comment here.